sharded_ttl.go

```diff
@@ -0,0 +1,139 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+// ShardedTTL holds the required variables to compose an in memory sharded cache system
+// which also provides expiring key mechanism
+type ShardedTTL struct {
+	// Mutex is used for handling the concurrent
+	// read/write requests for cache
+	sync.Mutex
+
+	// cache holds the cache data
+	cache *ShardedNoTS
+
+	// setAts holds the time that related item's set at, indexed by tenantId
+	setAts map[string]map[string]time.Time
+
+	// ttl is a duration for a cache key to expire
+	ttl time.Duration
+
+	// gcInterval is a duration for garbage collection
+	gcInterval time.Duration
+}
+
+// NewShardedWithTTL creates an in-memory sharded cache system
+// Which everytime will return the true values about a cache hit
+// and never will leak memory
+// ttl is used for expiration of a key from cache
+func NewShardedWithTTL(ttl time.Duration) *ShardedTTL {
+	return &ShardedTTL{
+		cache:  NewShardedNoTS(NewMemNoTSCache),
```

Comment:
could we make "NewMemNoTSCache" configurable just for the sake of extensibility?